torch/_inductor/codegen/common.py

```diff
@@ -1541,59 +1545,71 @@ def arg_to_bound(x):
-                    # Mixed negative and non-negative
-                    if var.bounds.upper >= 0:  # type: ignore[operator]
-                        lt = ops.lt(var, 0)
-                        stm = ops.where(lt, stm, var)
                     new_var = self.cse.generate(self.compute, stm, bounds=new_bounds)
-
+                    # Propagate the mask as mask propagation when using where is not correct
                     new_var.update_on_args("index_wrap", (var,), {})
                     var = new_var
 
-                if self.generate_assert(check):
-                    mask = self.load_mask(var)
-
-                    # An assertion line may have been written already, if so just
-                    # update the max size.
-                    map_key = (var, mask)
-                    existing_size, _ = self.indirect_max_sizes.get(
-                        map_key, (None, None)
+                sympy_var = parent_handler.indirect_indexing(var, size, check)
+                if generate_assert(check):
+                    assert_lower = not (var.bounds.lower >= 0)
+                    assert_upper = not isinstance(
+                        size, sympy.Number
+                    ) or not (  # value ranges cannot prove this
+                        var.bounds.upper < size
                     )
-                    if existing_size is not None:
-                        size = sympy.Min(size, existing_size)
-                    else:
-                        self.compute.writeline(
-                            IndirectAssertLine(
-                                "{assert_line}",
-                                self.indirect_assert,
-                                var,
-                                mask,
-                                self.indirect_max_sizes,
-                            )
-                        )
+                    CSEProxy.check_bounds_lazy(
+                        sympy_var, size, assert_lower, assert_upper
+                    )
+                return sympy_var
 
-                    self.indirect_max_sizes[map_key] = (size, self.index_to_str(size))
-                return parent_handler.indirect_indexing(var, size, check)
+            @staticmethod
+            def check_bounds_lazy(
+                expr: sympy.Expr, size: sympy.Expr, lower: bool, upper: bool
+            ):
+                """
+                How do we issue size checks:
+                - This function is called from either
+                  - CSEProxy.indirect_indexing
+                  - IndexPropagation.indirect_indexing
+                - This saves the info necessary to issue the check
+                - When a load or a store happens within codegen, it will call self.issue_check_bounds(buffer, index
+                - If the given index uses (structural equality) any of the expressions that
+                  need checking, we call self.check_bounds(buffer, ...) which will print the relevant line
```

Comment:
What does this do if the expression is simplified in the `load` or `store`. It sounds like we just wouldn't ever assert?

Reply:
We perform the same simplifications on the indexing and the expression itself, so I'd hope it still matches. I wanted to run CI on this to see if this is the case.

Reply:
This seems very fragile. Like what happens for a non-contiguous dimension:
```
index = i0 + 1
stride = 2
ops.load("buf0", 2*i0 + 2)
```
The pattern `i0 + 1` doesn't appear in the load index.

Reply:
what function are you looking at here? I have tested with indexing along a non-contiguous dimension and it generates everything as expected

Reply:
I'm only thinking of hypothetically at the moment, but I don't see how `(2*i0 + 2).has(i0 + 1)` could possibly work.

Reply:
Right. This function fails to generate the assert.
```python
@torch.compile
def fn(a):
    b = torch.arange(a.size(0) - 1, device=a.device) + 3
    return a[b]
```
Let me see how to strengthen this condition.

Reply:
Fixed via the introduction of a `Boxed` SymPy function. Also added comprehensive tests that cover this and other patterns by adding a `has_assert` option to the `index_propagation` tests.